Update for FFmpeg 3.5

walisser opened the pull request:
Summary of changes in v3.5 this accounts for:

- VOXWARE codec is removed
- AVCodecContext->codec_name is removed, replaced with avcodec_get_name()
- channel_layout is required to be set. Using av_get_default_channel_layout()
- some constants have AV_ prefix, using #define to account for old constant names

Additionally:
- fixes format string asserts in wxString::Printf (only tripped in debug build)

Assuming 2.2.x is the oldest we are supporting, tested with 2.2.9 and current git master. Versions in-between are assumed to work.

Paul-Licameli commented:
This looks like less work than I feared.  Thanks for the rapid work.

So, this can conditionally compile the right things either with the present contents of lib-src/ffmpeg, or at such time as we upgrade our lib-src.  Yes?

PRL

Paul-Licameli commented:
Darrell, please answer some questions.

Do you know how we can discover, since which version a file was in the interface?  I can't find such information here.  https://ffmpeg.org/doxygen/2.1/group__lavu__audio.html#gacb84f3e93a583e1f84a5283162a606a2

You add two FFMPEG_INITDYN calls.  So this may increase the possibilities that Audacity refuses to continue with an old FFmpeg library if it can't find either of the two new functions.  This did no happen with the version I have.  Is it likely with anybody else's versions?

You use the new functions in two places.  One is in import but only to get some information for logging messages.  The other sets a channel layout for export.  So, this is necessary to make export work in the new version.  It was not necessary with the old.  Could it cause any important change of behavior in the old?

I have tried this out on a simple mono track and confirmed that I can still export an .m4a and re-import it.  So, "First do no harm."

PRL

walisser commented:
On Fri, Jan 26, 2018 at 7:20 PM, Paul Licameli <notifications@github.com>
wrote:

> Darrell, please answer some questions.
>
> Do you know how we can discover, since which version a file was in the
> interface? I can't find such information here.
> https://ffmpeg.org/doxygen/2.1/group__lavu__audio.html#
> gacb84f3e93a583e1f84a5283162a606a2
>
Not sure how you would do that besides stepping through each version. There
is an "APIChanges" doc in FFmpeg but these aren't listed there.

https://ffmpeg.org/doxygen/2.0/group__lavu__audio.html#gacb84f3e93a583e1f84a5283162a606a2
https://ffmpeg.org/doxygen/2.2/group__lavu__audio.html#gacb84f3e93a583e1f84a5283162a606a2
​
https://ffmpeg.org/doxygen/2.4/group__channel__mask__c.html#gacb84f3e93a583e1f84a5283162a606a2
https://ffmpeg.org/doxygen/3.2/group__channel__mask__c.html#gacb84f3e93a583e1f84a5283162a606a2
https://ffmpeg.org/doxygen/3.4/group__channel__mask__c.html#gacb84f3e93a583e1f84a5283162a606a2

I'm not git expert, but I found this technique that seems suitable,
although a bit painful. Find the header with the symbol and use git log
with "show patch" option and release tag. But this won't notice a function
that moved from a different header in the past, in which case the search
could be expanded to more of the header files.

git log -p n1.2.9 libavutil/channel_layout.h | grep
av_get_default_channel_layout
+ *   av_get_default_channel_layout);
+uint64_t av_get_default_channel_layout(int nb_channels);

​Shows av_get_default_channel_layout() is present in the log, but not here:

git log -p n1.0.9 libavutil/channel_layout.h | grep
av_get_default_channel_layout
​

​Similarly:
​git log -p n0.8.9 libavcodec/avcodec.h | grep avcodec_get_name
​*nothing*
​

git log -p n0.9.4 libavcodec/avcodec.h | grep avcodec_get_name
   The table is queried using the avcodec_get_name function.
   If CONFIG_SMALL is true, the table is not compiled in; the
avcodec_get_name
+const char *avcodec_get_name(enum CodecID id);

You add two FFMPEG_INITDYN calls. So this may increase the possibilities
> that Audacity refuses to continue with an old FFmpeg library if it can't
> find either of the two new functions. This did no happen with the version I
> have. Is it likely with anybody else's versions?
>
​Based on my git search, not likely.​ These are both old APIs. Also I don't
see any breaking signature changes (the enum CodecID was renamed to
AVCodecID at some point, was the one change).

I think the best way to validate is to compile against the versions we need
to support and do an import/export check. This isn't a huge deal for me,
takes about 5 minutes per version to check. I don't know how far back I
needed to go, I believe cutoff is still 2.2 due to the Windows and OS X
builds using the lib-src version.

> You use the new functions in two places. One is in import but only to get
> some information for logging messages. The other sets a channel layout for
> export. So, this is necessary to make export work in the new version. It
> was not necessary with the old. Could it cause any important change of
> behavior in the old?
>
I think we can safely assume that channels == 1 or 2 will not cause any
change in behavior.

​It would be difficult to validate exactly what was happening before v3.5,
but a quick search shows av_get_default_channel_layout() was being used
heavily in the library and example programs even in v2.2

I have tried this out on a simple mono track and confirmed that I can still
> export an .m4a and re-import it. So, "First do no harm."
>
> PRL
>

Paul-Licameli commented:
On Sat, Jan 27, 2018 at 1:57 PM, Darrell Walisser <notifications@github.com>
wrote:

> On Fri, Jan 26, 2018 at 7:20 PM, Paul Licameli <notifications@github.com>
> wrote:
>
> > Darrell, please answer some questions.
> >
> > Do you know how we can discover, since which version a file was in the
> > interface? I can't find such information here.
> > https://ffmpeg.org/doxygen/2.1/group__lavu__audio.html#
> > gacb84f3e93a583e1f84a5283162a606a2
> >
> Not sure how you would do that besides stepping through each version. There
> is an "APIChanges" doc in FFmpeg but these aren't listed there.
>
> https://ffmpeg.org/doxygen/2.0/group__lavu__audio.html#
> gacb84f3e93a583e1f84a5283162a606a2
> https://ffmpeg.org/doxygen/2.2/group__lavu__audio.html#
> gacb84f3e93a583e1f84a5283162a606a2
> ​
> https://ffmpeg.org/doxygen/2.4/group__channel__mask__c.html#
> gacb84f3e93a583e1f84a5283162a606a2
> https://ffmpeg.org/doxygen/3.2/group__channel__mask__c.html#
> gacb84f3e93a583e1f84a5283162a606a2
> https://ffmpeg.org/doxygen/3.4/group__channel__mask__c.html#
> gacb84f3e93a583e1f84a5283162a606a2
>
> I'm not git expert, but I found this technique that seems suitable,
> although a bit painful. Find the header with the symbol and use git log
> with "show patch" option and release tag. But this won't notice a function
> that moved from a different header in the past, in which case the search
> could be expanded to more of the header files.
>
> git log -p n1.2.9 libavutil/channel_layout.h | grep
> av_get_default_channel_layout
> + * av_get_default_channel_layout);
> +uint64_t av_get_default_channel_layout(int nb_channels);
>
> ​Shows av_get_default_channel_layout() is present in the log, but not
> here:
>
> git log -p n1.0.9 libavutil/channel_layout.h | grep
> av_get_default_channel_layout
> ​
>
> ​Similarly:
> ​git log -p n0.8.9 libavcodec/avcodec.h | grep avcodec_get_name
> ​*nothing*
> ​
>

Thanks for this research.  So, we must go back in time to an FFmpeg version
from 2013, to lose either of those functions.  That's far enough I'll call
it safe.


>
> git log -p n0.9.4 libavcodec/avcodec.h | grep avcodec_get_name
> The table is queried using the avcodec_get_name function.
> If CONFIG_SMALL is true, the table is not compiled in; the
> avcodec_get_name
> +const char *avcodec_get_name(enum CodecID id);
>
> You add two FFMPEG_INITDYN calls. So this may increase the possibilities
> > that Audacity refuses to continue with an old FFmpeg library if it can't
> > find either of the two new functions. This did no happen with the
> version I
> > have. Is it likely with anybody else's versions?
> >
> ​Based on my git search, not likely.​ These are both old APIs. Also I don't
> see any breaking signature changes (the enum CodecID was renamed to
> AVCodecID at some point, was the one change).
>
> I think the best way to validate is to compile against the versions we need
> to support and do an import/export check. This isn't a huge deal for me,
> takes about 5 minutes per version to check. I don't know how far back I
> needed to go, I believe cutoff is still 2.2 due to the Windows and OS X
> builds using the lib-src version.
>
> > You use the new functions in two places. One is in import but only to get
> > some information for logging messages. The other sets a channel layout
> for
> > export. So, this is necessary to make export work in the new version. It
> > was not necessary with the old. Could it cause any important change of
> > behavior in the old?
> >
> I think we can safely assume that channels == 1 or 2 will not cause any
> change in behavior.
>
> ​It would be difficult to validate exactly what was happening before v3.5,
> but a quick search shows av_get_default_channel_layout() was being used
> heavily in the library and example programs even in v2.2
>

Do I understand that you discovered that export did not work in 3.5 unless
this call was added?  It's not needed for compilation.

Earlier versions may not have enforced it.  It appears to do no harm to add
the call.  Perhaps there is some subtle bug we haven't discovered that is
fixed by adding it.

I'm persuded to commit this now.  It should also satisfy the Debian people
that they won't need to wait for version 2.2.3 for a version that builds
with the coming FFmpeg.

Thank you Darrell.

PRL




>
> I have tried this out on a simple mono track and confirmed that I can still
> > export an .m4a and re-import it. So, "First do no harm."
> >
> > PRL
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/audacity/audacity/pull/249#issuecomment-361006723>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALITYZ-vOsq3l9mGvn9Zq9hkNSZD6I2uks5tO3GLgaJpZM4RvBOz>
> .
>

walisser commented on src/FFmpeg.h:
```diff
@@ -128,6 +128,13 @@ extern "C" {
       #endif
    #endif
+
+   #if LIBAVCODEC_VERSION_MAJOR < 58
```
Thank you for this catch, it's simple enough to test for existing defines and support everything.

Paul-Licameli replied:
Section A.12.3 of my good old K & R book (ANSI C version) says that when a a macro is expanded, occurrences of the same macro name in the resulting token sequence are not re-expanded.  So this will not produce undefined results from the preprocessor.  Rather AV_CODEC_FLAG_QSCALE will expand to CODEC_FLAG_QSCALE which will expand to AV_CODEC_FLAG_QSCALE again and then it stops.

Still, I'd agree we should not rely on that, and instead invent a third constant name to use in those few places in Audacity.

PRL

walisser commented:
I added a guard for those, and tested the even release versions
(2.2,2.4,2.6,2.8,3.0,3.2) with import/export test on each.

Rebased and pushed to the current branch. On hindsight I should have opened
a new PR?

On Sun, Jan 28, 2018 at 11:17 AM, Paul Licameli <notifications@github.com>
wrote:

> *@Paul-Licameli* commented on this pull request.
> ------------------------------
>
> In src/FFmpeg.h
> <https://github.com/audacity/audacity/pull/249#discussion_r164303854>:
>
> > @@ -128,6 +128,13 @@ extern "C" {
>        }
>        #endif
>     #endif
> +
> +   #if LIBAVCODEC_VERSION_MAJOR < 58
>
> Section A.12.3 of my good old K & R book (ANSI C version) says that when a
> a macro is expanded, occurrences of the same macro name in the resulting
> token sequence are not re-expanded. So this will not produce undefined
> results from the preprocessor. Rather AV_CODEC_FLAG_QSCALE will expand to
> CODEC_FLAG_QSCALE which will expand to AV_CODEC_FLAG_QSCALE again and then
> it stops.
>
> Still, I'd agree we should not rely on that, and instead invent a third
> constant name to use in those few places in Audacity.
>
> PRL
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/audacity/audacity/pull/249#discussion_r164303854>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA8r1TMkat989ND6iYxZjF0k2hzLU5sMks5tPJ2ZgaJpZM4RvBOz>
> .
>

Paul-Licameli commented:
No need for a new PR.  Thanks, I have merged it again.

mirh commented:
> I added a guard for those, and tested the even release versions (2.2,2.4,2.6,2.8,3.0,3.2) with import/export test on each.

Not sure what you tested, but I (along with [many other people](https://superuser.com/questions/1211211/import-ffmpeg-into-audacity)) cannot load any version newer than 2.2, or api 55 if you'd like
